src/coreclr/jit/fgopt.cpp

```diff
@@ -2354,12 +2354,12 @@ bool Compiler::fgOptimizeEmptyBlock(BasicBlock* block)
     {
         case BBJ_COND:
         case BBJ_SWITCH:
-        case BBJ_THROW:
 
             /* can never happen */
-            noway_assert(!"Conditional, switch, or throw block with empty body!");
+            noway_assert(!"Conditional or switch block with empty body!");
             break;
 
+        case BBJ_THROW:
```

Comment:
NOTE: `CORINFO_HELP_CHKCASTCLASS_SPECIAL` can be eliminated later (if its argument is null - it returns null) so we can get an empty `BBJ_THROW` which is ok. Repro:
```csharp
static string Test(object o1, object o2)
{
    if (o1 != null && o1 is string)
        return ((string)o1);
    return null;
}
```

Reply:
@AndyAyersMS so for:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static void Test(object o)
{
    if (o is string)
        Consume((string)o);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Consume(string str) { }
```
The first `o is string` cast is imported without any helper calls (we don't need to throw an exception) the second cast does emit a helper for `InvalidCastException` - but the first `is` makes it impossible to throw any exception basically, here is the full jitdump for this method from my branch (with the assert on): https://gist.github.com/EgorBo/21f81f1586f8d90de0e0480d42b2b87b

`optAssertionProp_Call`  then just replaces that `call ` with its second argument (local), see https://github.com/dotnet/runtime/blob/1d36a90ad6fd2a7985bc10f7b2c617caaf8844d5/src/coreclr/jit/assertionprop.cpp#L3970-L3998 and that local then is removed.

so we ended up with an empty block which was imported as `BBJ_THROW`. If I remove that assert - that block is then collected (by optOptimizeLayout?)

Reply:
Ah right, I forgot we can't just get rid of the redundant checks in the importer. Thanks.